enterprise/backend/src/metabase_enterprise/metabot_v3/tools/generate_insights.clj

```diff
@@ -0,0 +1,16 @@
+(ns metabase-enterprise.metabot-v3.tools.generate-insights
+  (:require
+   [metabase-enterprise.metabot-v3.tools.interface :as metabot-v3.tools.interface]
+   [metabase.public-settings :as public-settings]
+   [metabase.util.malli :as mu]))
+
+(mu/defmethod metabot-v3.tools.interface/*invoke-tool* :metabot.tool/generate-insights
+  [_tool-name {what-for :for :as _arguments} context]
+  (let [[k id] (some #(find what-for %) [:metric_id :table_id :report_id :query_id])
+        entity-type (case k
+                      :metric_id "metric"
+                      :table_id "table"
+                      :report_id "card"
+                      :query_id (throw (ex-info "Unhandled" {})))]
```

Comment:
This should be handled by using `"adhoc"` as entity type and the base64 encoded `dataset_query` as ID.

According to my current [vision](https://www.notion.so/Metabot-toolbox-v1-15169354c90180cf92edd3a32f8fc62c?d=15969354c90180a998d4001caa436381&pvs=4#15169354c9018014a350ff62edc9e914) with some rudiments implemented in #51315, is that you can find the query by this ID in what I call the full history.